zkevm-circuits/src/test_util.rs

```diff
@@ -42,130 +41,201 @@ impl Default for BytecodeTestConfig {
-        let power_of_randomness = state_circuit.instance();
-        let prover = MockProver::<Fr>::run(18, &state_circuit, power_of_randomness).unwrap();
-        // Skip verification of Start rows to accelerate testing
-        let non_start_rows_len = state_circuit
-            .rows
-            .iter()
-            .filter(|rw| !matches!(rw, Rw::Start { .. }))
-            .count();
-        prover.verify_at_rows(
-            N_ROWS - non_start_rows_len..N_ROWS,
-            N_ROWS - non_start_rows_len..N_ROWS,
-        )?
+    /// Allows to provide checks different than the default ones for the EVM
+    /// Circuit verification.
+    pub fn evm_checks(mut self, evm_checks: Box<dyn Fn(MockProver<Fr>)>) -> Self {
+        self.evm_checks = Some(evm_checks);
+        self
     }
 
-    Ok(())
+    /// Allows to provide modifier functions for the [`Block`] that will be
+    /// generated within this builder.
+    ///
+    /// That prevents to a lot of thests the need to build the block outside of
+    /// the builder because they need to modify something particular.
+    pub fn block_modifier(mut self, modifier: Box<dyn Fn(&mut Block<Fr>)>) -> Self {
+        self.block_modifiers.push(modifier);
+        self
+    }
 }
 
-/// generate rand tx for pi circuit
-pub fn rand_tx<R: Rng + CryptoRng>(mut rng: R, chain_id: u64, has_calldata: bool) -> Transaction {
-    let wallet0 = LocalWallet::new(&mut rng).with_chain_id(chain_id);
-    let wallet1 = LocalWallet::new(&mut rng).with_chain_id(chain_id);
-    let from = wallet0.address();
-    let to = wallet1.address();
-    let data = if has_calldata {
-        let mut data = b"helloworld".to_vec();
-        data[0] = 0;
-        data[2] = 0;
-        data
-    } else {
-        vec![]
-    };
-    let tx = TransactionRequest::new()
-        .chain_id(chain_id)
-        .from(from)
-        .to(to)
-        .nonce(3)
-        .value(1000)
-        .data(data)
-        .gas(500_000)
-        .gas_price(1234);
-    let sig = wallet0.sign_transaction_sync(&tx.clone().into());
-    let to = tx.to.map(|to| match to {
-        NameOrAddress::Address(a) => a,
-        _ => unreachable!(),
-    });
-    Transaction {
-        from: tx.from.unwrap(),
-        to,
-        gas_limit: tx.gas.unwrap(),
-        gas_price: tx.gas_price.unwrap(),
-        value: tx.value.unwrap(),
-        call_data: tx.data.unwrap(),
-        nonce: tx.nonce.unwrap(),
-        v: sig.v,
-        r: sig.r,
-        s: sig.s,
-        ..Transaction::default()
+impl<const NACC: usize, const NTX: usize> CircuitTestBuilder<NACC, NTX> {
+    /// Triggers the `CircuitTestBuilder` to convert the [`TestContext`] if any,
+    /// into a [`Block`] and apply the default or provided block_modifiers or
+    /// circuit checks to the provers generated for the State and EVM circuits.
+    pub fn run(self) {
+        let block: Block<Fr> = if self.block.is_some() {
+            self.block.unwrap()
```

Comment:
suggestion to detect missuses of the `CircuitTestBuilder`.  When a block is set, it already contains the `block.circuits_params`.  So the user shouldn't set another circuits_params because it will be ignored.  So in this case:
```rust
assert!(self.circuit_params.is_none(), "circuit_params already provided in the block");
```

Reply:
That's a good point! Will do the same for `block` and `ctx` as both shouldn't be provided at the same time to the builder.

Reply:
Added in 4565a05e